### DISCUSSION TITLE
流式数据插入时，每次100条左右的向量，几万条数据就要占用几个G的内存，请问怎么回事，谢谢

### DISCUSSION THREAD

**Discussion by liuyichen2013** [milvus-io/milvus#15205] [Q&A and General discussion]
流式数据插入时，每次100条左右的向量，几万条数据就要占用几个G的内存，请问怎么回事，谢谢

**Comment by yhmo** [comment]
Which version of your milvus?

**Reply by liuyichen2013** [reply]
rc8 和pre-ga都有这种情况。
在插入向量的同时，将向量load到内存，就会出现内存占用过高的情况。
如果将所有的向量都插入完成后，再load到内存，则是正常的。

谢谢

**Reply by yhmo** [reply]
每次插入100条向量之后，有没有调用过譬如collection.num_entities()这样的操作？

**Reply by liuyichen2013** [reply]
调用或者不调用collection.num_entities()都一样的现象。

**Reply by yhmo** [reply]
querynode的机制是按照shard number创建等量的growing segment在内存里，每个大小是32768行空数据，这里会有一个初始的占用量。在flush之后，产生100行数据的sealed segment并持久化，然后load，这100行数据加载到querynode的内存。所以理论上内存的占用量是+100 +100 +100这么涨的。如果实际表现不一样，有可能是个bug。麻烦你在github issue里创建一个issue，详细描述一下，包括部署方式，config有无改动，客户端脚本(伪码也行)。

**Reply by liuyichen2013** [reply]
我是从kafka获取向量，然后每分钟写入一次向量库，每次大概100条向量，代码如下：
`#自动生成主键
        default_fields = [
            FieldSchema(name="id", dtype=DataType.INT64, is_primary=True, auto_id=True),
            FieldSchema(name="relate_id", dtype=DataType.INT64),
            FieldSchema(name="ts", dtype=DataType.INT64),
            FieldSchema(name="ft", dtype=d_type, dim=dim)
        ]
        default_schema = CollectionSchema(fields=default_fields, description="milvus collection")

        if utility.has_collection(collection_name):
            collection = Collection(collection_name)
            logger.info("collection:" + collection_name + ' already exists')
        else:
            collection = Collection(name=collection_name, schema=default_schema)
            default_index = {"index_type": "IVF_FLAT", "params": {"nlist": milvusnList}, "metric_type": "L2"}
            collection.create_index(field_name="ft", index_params=default_index)
            collection.load()
            logger.info("create collection :" + collection_name + ' sucessed')

        entities = [relate_id, ts, vectors]
        collection.insert(entities)`

**Reply by yhmo** [reply]
还缺少部署的方式以及配置，另外向量的维度也要告知。最好能建一个issue，issue的好处是可以assign到相关人员，这样可以先让测试团队验证，然后转交给开发人员。

**Comment by liuyichen2013** [comment]
rc8 和pre-ga都有这种情况。
在插入向量的同时，将向量load到内存，就会出现内存占用过高的情况。
如果将所有的向量都插入完成后，再load到内存，则是正常的。

谢谢

**Comment by xiaofan-luan** [comment]
这个是正常情况，go本身也会有很多中间状态的开销，流式插入的话链路会比较长，带来一些overhead。我们推荐分布式Milvus最小规模是4c8G以上，分布式版本的话8c32G以上

**Reply by yanliang567** [reply]
分布式版本最好是16c64G以上，8c32G还是比较紧张

**Reply by liuyichen2013** [reply]
这种情况，即使64G内存，如果流式数据有上百万的话，内存也是不够的！

**Reply by yhmo** [reply]
在初始阶段内存消耗曲线比较陡，后面应该不这么陡的，@liuyanfeng1981 你可以观察一下。
